### PR TITLE
Hooks to use netket with single precision

### PR DESCRIPTION
**Pull request by PhilipVinc**
@wdphy16 @inailuig are there some things that you hack/patch or that annoy you that we need to be able to fully run without x64 enabled?

### CODE REVIEW

**Comment by wdphy16** [comment]
Looks great! Actually I caught exactly those two things some time ago when I wanted to run NetKet fully in x32 ( https://github.com/wdphy16/netket/commit/fb8a9b874e3bc7bf462bfc73f93ed3bc8cfa96b0 , https://github.com/wdphy16/netket/commit/b37d2fe856821dfc1f3952587d6ba67712866a88 )

Maybe it's good to have CI tests with x64 disabled, but we need to revise a lot of numerical tolerances when comparing numbers...

**Comment by PhilipVinc** [comment]
> Maybe it's good to have CI tests with x64 disabled, but we need to revise a lot of numerical tolerances when comparing numbers...

that's a very good idea, actually. I'll try to get this sorted out at some point

> Looks great! Actually I caught exactly those two things some time ago when I wanted to run NetKet fully in x32

Nice! Did you catch anything else? Am I missing something else to address this?

> We also need to mute warnings like UserWarning: Explicitly requested dtype <class 'jax.numpy.float64'>, e.g., just type import netket.experimental as nkx and there are a lot of them.

Yes, that's why I'm trying to fix all operator promotion in operators. If I fix that logic, then I just need to make sure to fix the constructors, and model definitions. 

Actually, I am unsure if we should have a flag `NETKET_DEFAULT_PRECISION` which defaults to `64 bit` but can be switched to `32 bit`. Then, we should just make sure that our models use that precision by default.

Ideally I'd like to run in 64 bit mode but have models and gradients computed in 32 bit, and just have the local energy pooling run in 64 bit...

what do you think @wdphy16 about this?

**Comment by wdphy16** [comment]
> Nice! Did you catch anything else? Am I missing something else to address this?

I didn't catch anything else yet, and those two are what prevents the code from running. Maybe we also need to check the type stability in all tests

> Actually, I am unsure if we should have a flag `NETKET_DEFAULT_PRECISION` which defaults to `64 bit` but can be switched to `32 bit`. Then, we should just make sure that our models use that precision by default.

I guess it would not be too hard using our infrastructure of flags. Recently I'm not dealing with this kind of precision problems but you can try it

**Comment by PhilipVinc** [comment]
@wdphy16 I think the PR as of now should greatly reduce the number of warnings it throws because of compatibility. 
Can you try giving me some codes that raise more warnings?

File: netket/operator/_local_operator/compile_helpers.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ def pack_internals(
     local_states = np.full(
         (n_operators, max_acting_on_sz, max_local_hilbert_size), np.nan
     )
-    basis = np.full((n_operators, max_acting_on_sz), 1e10, dtype=np.int64)
+    basis = np.full((n_operators, max_acting_on_sz), 0x7FFFFFFF, dtype=np.int64)
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/operator/_local_operator/compile_helpers.py:75]: It's a subliminal satanic message invented by @wdphy16

[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/operator/_local_operator/compile_helpers.py:75]: on a serious note, it's the largest integer you can represent on an `int32`. 
This is just a flag value used to signal that some entries are unset, and using this value make it not warn if we use int32 .

**Review by wdphy16 — APPROVED** [review]
Looks good to me, but can you see what tests failed? It seems I can't see that
